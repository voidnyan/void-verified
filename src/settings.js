import { defaultSettings } from "./defaultSettings";
import { ColorFunctions } from "./colorFunctions";
import { AnilistAPI } from "./api/anilistAPI";

class Option {
	value;
	defaultValue;
	description;
	category;
	constructor(option) {
		this.defaultValue = option.defaultValue;
		this.description = option.description;
		this.category = option.category;
	}

	getValue() {
		if (this.value === "") {
			return this.defaultValue;
		}
		return this.value ?? this.defaultValue;
	}
}

export class Settings {
	localStorageUsers = "void-verified-users";
	localStorageSettings = "void-verified-settings";
	version = GM_info.script.version;

	verifiedUsers = [];

	options = {};

	constructor() {
		this.verifiedUsers =
			JSON.parse(localStorage.getItem(this.localStorageUsers)) ?? [];

		const settingsInLocalStorage =
			JSON.parse(localStorage.getItem(this.localStorageSettings)) ?? {};

		for (const [key, value] of Object.entries(defaultSettings)) {
			this.options[key] = new Option(value);
		}

		for (const [key, value] of Object.entries(settingsInLocalStorage)) {
			if (!this.options[key]) {
				continue;
			}
			this.options[key].value = value.value;
		}
	}

	verifyUser(username) {
		if (
			this.verifiedUsers.find(
				(user) => user.username.toLowerCase() === username.toLowerCase()
			)
		) {
			return;
		}

		this.verifiedUsers.push({ username });
		localStorage.setItem(
			this.localStorageUsers,
			JSON.stringify(this.verifiedUsers)
		);

		const anilistAPI = new AnilistAPI(this);
		anilistAPI.queryUserData();
	}

	updateUserOption(username, key, value) {
		this.verifiedUsers = this.verifiedUsers.map((u) =>
			u.username === username
				? {
						...u,
						[key]: value,
				  }
				: u
		);
		localStorage.setItem(
			this.localStorageUsers,
			JSON.stringify(this.verifiedUsers)
		);
	}

	updateUserFromApi(user, apiUser) {
		const newUser = this.#mapApiUser(user, apiUser);
		this.verifiedUsers = this.verifiedUsers.map((u) =>
			u.username.toLowerCase() === user.username.toLowerCase()
				? newUser
				: u
		);

		localStorage.setItem(
			this.localStorageUsers,
			JSON.stringify(this.verifiedUsers)
		);
	}

	#mapApiUser(user, apiUser) {
		let userObject = { ...user };

		userObject.color = this.#handleAnilistColor(
			apiUser.options.profileColor
		);

		userObject.username = apiUser.name;
		userObject.avatar = apiUser.avatar.large;
		userObject.lastFetch = new Date();

		return userObject;
	}

	removeUser(username) {
		this.verifiedUsers = this.verifiedUsers.filter(
			(user) => user.username !== username
		);
		localStorage.setItem(
			this.localStorageUsers,
			JSON.stringify(this.verifiedUsers)
		);
	}

	saveSettingToLocalStorage(key, value) {
		let localSettings = JSON.parse(
			localStorage.getItem(this.localStorageSettings)
		);

		this.options[key].value = value;

		if (localSettings === null) {
			const settings = {
				[key]: value,
			};
			localStorage.setItem(
				this.localStorageSettings,
				JSON.stringify(settings)
			);
			return;
		}

		localSettings[key] = { value };
		localStorage.setItem(
			this.localStorageSettings,
			JSON.stringify(localSettings)
		);
	}

	#defaultColors = [
		"gray",
		"blue",
		"purple",
		"green",
		"orange",
		"red",
		"pink",
	];

	#defaultColorRgb = {
		gray: "103, 123, 148",
		blue: "61, 180, 242",
		purple: "192, 99, 255",
		green: "76, 202, 81",
		orange: "239, 136, 26",
		red: "225, 51, 51",
		pink: "252, 157, 214",
	};

	#handleAnilistColor(color) {
		if (this.#defaultColors.includes(color)) {
			return this.#defaultColorRgb[color];
		}

		return ColorFunctions.hexToRgb(color);
	}
}
