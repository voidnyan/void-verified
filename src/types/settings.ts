export interface ISettings {
	options: IOptions;
	version: string;
	auth: {};
	anilistUser: string;
	isAuthorized: () => boolean;
}

export interface IOption {
	value?: string | number | boolean;
	defaultValue: string | number | boolean;
	description: string;
	category: string;
	authRequired: boolean;
	getValue: () => string | number | boolean;
}

export interface IOptions {
	copyColorFromProfile: IOption;
	moveSubscribeButtons: IOption;
	hideLikeCount: IOption;
	enabledForUsername: IOption;
	enabledForProfileName: IOption;
	defaultSign: IOption;
	highlightEnabled: IOption;
	highlightEnabledForReplies: IOption;
	highlightSize: IOption;
	colorUserActivity: IOption;
	colorUserReplies: IOption;
	useDefaultHighlightColor: IOption;
	defaultHighlightColor: IOption;
	globalCssEnabled: IOption;
	globalCssAutoDisable: IOption;
	profileCssEnabled: IOption;
	activityCssEnabled: IOption;
	onlyLoadCssFromVerifiedUser: IOption;
	layoutDesignerEnabled: IOption;
	quickAccessNotificationsEnabled: IOption;
	quickAccessEnabled: IOption;
	quickAccessBadge: IOption;
	quickAccessTimer: IOption;
	pasteEnabled: IOption;
	pasteWrapImagesWithLink: IOption;
	pasteImageWidth: IOption;
	pasteImagesToHostService: IOption;
	toasterEnabled: IOption;
	removeAnilistBlanks: IOption;
	gifKeyboardEnabled: IOption;
	gifKeyboardLikeButtonsEnabled: IOption;
	changeLogEnabled: IOption;
	selfMessageEnabled: IOption;
	hideMessagesFromListFeed: IOption;
	csspyEnabled: IOption;
	replyActivityUpdate: IOption;
	markdownHotkeys: IOption
	// useElevatedFetch: IOption;
}
