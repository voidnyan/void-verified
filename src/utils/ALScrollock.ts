export class ALScrollock {
	static lock() {
		document.body.classList.add("scroll-lock");
	}

	static unlock(){
		document.body.classList.remove("scroll-lock");
	}
}
