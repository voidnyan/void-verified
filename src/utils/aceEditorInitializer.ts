export class AceEditorInitializer {
	static initializeEditor(id: string, value: string) {
		const siteTheme = localStorage.getItem("site-theme");
		let theme = "monokai";
		switch (siteTheme) {
			case "dark":
			case "system":
				theme = "monokai";
				break;
			case "contrast":
			default:
				theme = "dawn";
		}
		// @ts-ignore
		const editor = ace.edit(id, {
			mode: "ace/mode/css",
			theme: `ace/theme/${theme}`,
			value: value
		});

		editor.setKeyboardHandler("ace/keyboard/vscode");

		editor.setOptions({
			useWorker: false,
			enableBasicAutocompletion: true,
			highlightSelectedWord: true,
			copyWithEmptySelection: true,
			scrollPastEnd: true,
			showPrintMargin: false
		});
	}

	static addChangeHandler(id: string, callback: (value: string) => void) {
		// @ts-ignore
		const editor = ace.edit(id)
		editor.on("change", () => {
			callback(editor.getValue());
		})
	}
}
