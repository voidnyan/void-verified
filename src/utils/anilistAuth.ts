import {imageHosts, ImageHostService} from "../api/imageHostConfiguration";
import {ImgurAPI} from "../api/imgurAPI";
import {DOM} from "./DOM";
import {StaticSettings} from "./staticSettings";
import {ButtonComponent} from "../components/ButtonComponent";
import {VoidApi} from "../api/voidApi";

export class AnilistAuth {
	private static localStorageAuth = "void-verified-auth";
	static token: string = null;
	static expires: Date = null;

	static name: string;
	static id: number;

	private static settingsContainer = DOM.createDiv();

	static initialize() {
		const auth = JSON.parse(localStorage.getItem(this.localStorageAuth)) ?? null;
		if (auth) {
			this.token = auth.token;
			this.expires = new Date(auth.expires);
		}


		const alAuth = JSON.parse(localStorage.getItem("auth"));
		this.name = alAuth?.name;
		this.id = alAuth?.id;
	}

	static checkAuthFromUrl() {
		const hash = window.location.hash.substring(1);
		if (!hash) {
			return;
		}

		const [path, token, type, expiress] = hash.split("&");

		if (path === "void_imgur") {
			const imgurConfig =
				ImageHostService.getImageHostConfiguration(
					imageHosts.imgur,
				);
			new ImgurAPI(imgurConfig).handleAuth();
		}

		if (path === "void_api_auth") {
			VoidApi.token = token;
			localStorage.setItem("void-verified-api-token", token);
		}

		if (path === "void_auth") {
			const expiresDate = new Date(
				new Date().getTime() + Number(expiress.split("=")[1]) * 1000,
			);

			this.saveAuthToken({
				token: token.split("=")[1],
				expires: expiresDate,
			});
		}

		window.history.replaceState(
			null,
			"",
			"https://anilist.co/settings/developer",
		);
	}

	static saveAuthToken(tokenObject: {token: string, expires: Date}) {
		this.token = tokenObject.token;
		this.expires = tokenObject.expires;
		localStorage.setItem(
			this.localStorageAuth,
			JSON.stringify(tokenObject),
		);
	}

	static createSettings() {
		this.renderSettings();
		return this.settingsContainer;
	}

	private static renderSettings() {
		this.settingsContainer.replaceChildren();
		const isAuthenticated =
			this.token !== null &&
			new Date(this.expires) > new Date();

		const clientId = 15519;

		const header = DOM.create("h3", null, "Authorize (AniList API)");
		const description = DOM.create(
			"p",
			null,
			"Some features of VoidVerified might need your access token to work correctly or fully. Below is a list of features using your access token. If you do not wish to use any of these features, you do not need to authorize. If revoking authentication, be sure to revoke VoidVerified from Anilist Apps as well.",
		);

		const list = DOM.create("ul");
		for (const option of Object.values(StaticSettings.options).filter(
			(o) => o.authRequired,
		)) {
			list.append(DOM.create("li", null, option.description));
		}

		// DOM.create uses Vue router so don't use that here
		const authLink = document.createElement("a");
		authLink.classList.add("void-button");
		authLink.append("Authorize VoidVerified");
		authLink.setAttribute(
			"href",
			`https://anilist.co/api/v2/oauth/authorize?client_id=${clientId}&response_type=token`,
		);

		const removeAuthButton = DOM.create(
			"button",
			null,
			"Revoke auth token",
		);
		removeAuthButton.classList.add("button");
		removeAuthButton.addEventListener("click", () => {
			this.removeAuthToken();
			this.renderSettings();
		});

		this.settingsContainer.append(header);
		this.settingsContainer.append(description);
		this.settingsContainer.append(list);
		this.settingsContainer.append(
			!isAuthenticated ? authLink : removeAuthButton,
		);


		this.settingsContainer.append(VoidApi.createSettings());
	}

	static 	removeAuthToken() {
		this.token = null;
		this.expires = null;
		localStorage.removeItem(this.localStorageAuth);
	}
}
