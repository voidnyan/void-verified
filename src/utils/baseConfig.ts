export class BaseConfig {
	configInLocalStorage: string;
	constructor(configInLocalStorage: string) {
		this.configInLocalStorage = configInLocalStorage;
	}
	save() {
		localStorage.setItem(this.configInLocalStorage, JSON.stringify(this));
	}

}
