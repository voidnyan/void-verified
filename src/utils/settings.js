import { categories, defaultSettings } from "../assets/defaultSettings";
import { ColorFunctions } from "./colorFunctions";
import { AnilistAPI } from "../api/anilistAPI";
import { Toaster } from "./toaster";

class Option {
	value;
	defaultValue;
	description;
	category;
	authRequired;
	constructor(option) {
		this.defaultValue = option.defaultValue;
		this.description = option.description;
		this.category = option.category ?? categories.misc;
		this.authRequired = option.authRequired;
	}

	getValue() {
		if (this.value === "") {
			return this.defaultValue;
		}
		return this.value ?? this.defaultValue;
	}
}

export class Settings {
	localStorageUsers = "void-verified-users";
	localStorageSettings = "void-verified-settings";
	localStorageAuth = "void-verified-auth";
	version;
	auth = null;
	anilistUser;

	verifiedUsers = [];

	options = {};

	constructor() {
		this.version = GM_info.script.version;
		this.verifiedUsers =
			JSON.parse(localStorage.getItem(this.localStorageUsers)) ?? [];

		// add change to test pr unit tests

		const settingsInLocalStorage =
			JSON.parse(localStorage.getItem(this.localStorageSettings)) ?? {};

		for (const [key, value] of Object.entries(defaultSettings)) {
			this.options[key] = new Option(value);
		}

		for (const [key, value] of Object.entries(settingsInLocalStorage)) {
			if (!this.options[key]) {
				continue;
			}
			this.options[key].value = value.value;
		}

		this.auth =
			JSON.parse(localStorage.getItem(this.localStorageAuth)) ?? null;

		const auth = JSON.parse(localStorage.getItem("auth"));
		this.anilistUser = auth?.name;
	}

	async verifyUser(username) {
		if (
			this.verifiedUsers.find(
				(user) => user.username.toLowerCase() === username.toLowerCase()
			)
		) {
			return;
		}

		this.verifiedUsers.push({ username });
		localStorage.setItem(
			this.localStorageUsers,
			JSON.stringify(this.verifiedUsers)
		);

		try {
			Toaster.debug(`Querying ${username}.`);
			const anilistAPI = new AnilistAPI(this);
			const user = await anilistAPI.queryUser(username);
			this.updateUserFromApi(user);
		} catch (error) {
			Toaster.error("Failed to query new user.");
			console.error(error);
		}
	}

	getUser(username) {
		return this.verifiedUsers.find((user) => user.username === username);
	}

	isVerified(username) {
		return this.verifiedUsers.some((user) => user.username === username);
	}

	updateUserOption(username, key, value) {
		this.verifiedUsers = this.verifiedUsers.map((u) =>
			u.username === username
				? {
						...u,
						[key]: value,
				  }
				: u
		);
		localStorage.setItem(
			this.localStorageUsers,
			JSON.stringify(this.verifiedUsers)
		);
	}

	updateUserFromApi(apiUser) {
		let user = this.#findVerifiedUser(apiUser);

		if (!user) {
			return;
		}

		const newUser = this.#mapApiUser(user, apiUser);
		this.#mapVerifiedUsers(newUser);

		localStorage.setItem(
			this.localStorageUsers,
			JSON.stringify(this.verifiedUsers)
		);
	}

	#findVerifiedUser(apiUser) {
		let user = this.verifiedUsers.find((u) => u.id && u.id === apiUser.id);

		if (user) {
			return user;
		}

		return this.verifiedUsers.find(
			(u) => u.username.toLowerCase() === apiUser.name.toLowerCase()
		);
	}

	#mapVerifiedUsers(newUser) {
		if (this.verifiedUsers.find((u) => u.id && u.id === newUser.id)) {
			this.verifiedUsers = this.verifiedUsers.map((u) =>
				u.id === newUser.id ? newUser : u
			);
			return;
		}
		this.verifiedUsers = this.verifiedUsers.map((u) =>
			u.username.toLowerCase() === newUser.username.toLowerCase()
				? newUser
				: u
		);
	}

	#mapApiUser(user, apiUser) {
		let userObject = { ...user };

		userObject.color = ColorFunctions.handleAnilistColor(
			apiUser.options.profileColor
		);

		userObject.username = apiUser.name;
		userObject.avatar = apiUser.avatar.large;
		userObject.banner = apiUser.bannerImage;
		userObject.id = apiUser.id;
		userObject.lastFetch = new Date();

		if (this.options.quickAccessBadge.getValue() || user.quickAccessBadge) {
			if (
				(user.avatar && user.avatar !== userObject.avatar) ||
				(user.color && user.color !== userObject.color) ||
				(user.banner && user.banner !== userObject.banner) ||
				(user.username &&
					user.username.toLowerCase() !==
						userObject.username.toLowerCase())
			) {
				userObject.quickAccessBadgeDisplay = true;
			}
		}

		return userObject;
	}

	saveAuthToken(tokenObject) {
		this.auth = tokenObject;
		localStorage.setItem(
			this.localStorageAuth,
			JSON.stringify(tokenObject)
		);
	}

	removeAuthToken() {
		this.auth = null;
		localStorage.removeItem(this.localStorageAuth);
	}

	removeUser(username) {
		this.verifiedUsers = this.verifiedUsers.filter(
			(user) => user.username !== username
		);
		localStorage.setItem(
			this.localStorageUsers,
			JSON.stringify(this.verifiedUsers)
		);
	}

	saveSettingToLocalStorage(key, value) {
		let localSettings = JSON.parse(
			localStorage.getItem(this.localStorageSettings)
		);

		this.options[key].value = value;

		if (localSettings === null) {
			const settings = {
				[key]: value,
			};
			localStorage.setItem(
				this.localStorageSettings,
				JSON.stringify(settings)
			);
			return;
		}

		localSettings[key] = { value };
		localStorage.setItem(
			this.localStorageSettings,
			JSON.stringify(localSettings)
		);
	}
}
