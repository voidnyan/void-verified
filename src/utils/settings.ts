import { IOptions, ISettings } from "../types/settings";
import {StaticSettings} from "./staticSettings";
import {LocalStorageKeys} from "../assets/localStorageKeys";
import {VerifiedUsers} from "./verifiedUsers";
import {AnilistAuth} from "./anilistAuth";

export class Settings implements ISettings {
	localStorageSettings = LocalStorageKeys.settings;
	localStorageAuth = "void-verified-auth";
	version;
	auth = {
		token: AnilistAuth.token,
		expires: AnilistAuth.expires
	};
	anilistUser: string;
	userId;

	verifiedUsers = VerifiedUsers.users;

	options: IOptions = StaticSettings.options;

	constructor() {
		this.version = GM_info.script.version;

	}

	isAuthorized() {
		const isAuthorized = this.auth?.token != null;
		return isAuthorized;
	}

	updateUserOption(username, key, value) {
		VerifiedUsers.updateUserOption(username, key, value);
	}

	removeAuthToken() {
		this.auth = null;
		localStorage.removeItem(this.localStorageAuth);
	}

	saveSettingToLocalStorage(key: string, value: string | boolean) {
		let localSettings = JSON.parse(
			localStorage.getItem(this.localStorageSettings),
		);

		this.options[key].value = value;

		if (localSettings === null) {
			const settings = {
				[key]: value,
			};
			localStorage.setItem(
				this.localStorageSettings,
				JSON.stringify(settings),
			);
			return;
		}

		localSettings[key] = { value };
		localStorage.setItem(
			this.localStorageSettings,
			JSON.stringify(localSettings),
		);
	}
}
