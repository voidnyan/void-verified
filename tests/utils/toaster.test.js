const {
	Toaster,
	ToasterConfig,
	toastLevels,
	toastTypes,
} = require("../../src/utils/toaster");
import { Settings } from "../../src/utils/settings";

beforeAll(() => {
	jest.spyOn(global.Math, "random").mockReturnValue(0.12352);
});

afterAll(() => {
	jest.spyOn(global.Math, "random").mockRestore();
});

const renderSettingsUi = jest.fn();

describe("Toaster", () => {
	beforeEach(() => {
		document.body.replaceChildren();
	});

	test.each([
		[toastLevels.info, toastTypes.info, true],
		[toastLevels.info, toastTypes.success, true],
		[toastLevels.info, toastTypes.warning, true],
		[toastLevels.info, toastTypes.error, true],
		[toastLevels.success, toastTypes.info, false],
		[toastLevels.success, toastTypes.success, true],
		[toastLevels.success, toastTypes.warning, true],
		[toastLevels.success, toastTypes.error, true],
		[toastLevels.warning, toastTypes.info, false],
		[toastLevels.warning, toastTypes.success, false],
		[toastLevels.warning, toastTypes.warning, true],
		[toastLevels.warning, toastTypes.error, true],
		[toastLevels.error, toastTypes.info, false],
		[toastLevels.error, toastTypes.success, false],
		[toastLevels.error, toastTypes.warning, false],
		[toastLevels.error, toastTypes.error, true],
	])(
		"toast level %p should render %p message %p",
		(toastLevel, toastType, shouldRender) => {
			configureToaster(toastLevel);
			raiseToast(toastType, toastLevel);

			const toast = document.querySelector(".void-toast");
			if (shouldRender) {
				expect(toast).not.toBeNull();
			} else {
				expect(toast).toBeNull();
			}
		}
	);

	it("renders settings", () => {
		configureToaster(0);

		const container = Toaster.renderSettings(renderSettingsUi);
		expect(container).toMatchSnapshot();
	});
});

const configureToaster = (toastLevel) => {
	const settings = new Settings();
	settings.options.toasterEnabled.value = true;
	const toasterConfig = new ToasterConfig({ toastLevel });
	localStorage.setItem(
		"void-verified-toaster-config",
		JSON.stringify(toasterConfig)
	);
	Toaster.initializeToaster(settings);
};

const raiseToast = (toastType, toastLevel) => {
	switch (toastType) {
		case toastTypes.info:
			Toaster.debug(`toastLevel ${toastLevel}`);
			break;
		case toastTypes.success:
			Toaster.success(`toastLevel ${toastLevel}`);
			break;
		case toastTypes.warning:
			Toaster.warning(`toastLevel ${toastLevel}`);
			break;
		case toastTypes.error:
			Toaster.error(`toastLevel ${toastLevel}`);
			break;
	}
};
