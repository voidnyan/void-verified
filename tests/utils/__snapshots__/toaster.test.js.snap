// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Toaster Settings renders settings 1`] = `
<div>
  <h3>
    Configure Toasts
  </h3>
  <p>
    Toasts are notifications that pop up in the corner of your screen when things are happening.
  </p>
  <div
    class="void-label-container"
  >
    <label
      class="void-label-span"
      for="0.12352"
    >
      Toast level
    </label>
    <div
      class="void-select"
      id="0.12352"
    >
      <div
        class="void-option active"
      >
        info
      </div>
      <div
        class="void-option"
      >
        success
      </div>
      <div
        class="void-option"
      >
        warning
      </div>
      <div
        class="void-option"
      >
        error
      </div>
    </div>
  </div>
  <div
    class="void-label-container"
  >
    <label
      class="void-label-span"
      for="0.12352"
    >
      Toast location
    </label>
    <div
      class="void-select"
      id="0.12352"
    >
      <div
        class="void-option"
      >
        top-left
      </div>
      <div
        class="void-option"
      >
        top-right
      </div>
      <div
        class="void-option active"
      >
        bottom-left
      </div>
      <div
        class="void-option"
      >
        bottom-right
      </div>
    </div>
  </div>
  <div
    class="void-label-container"
  >
    <label
      class="void-label-span"
      for="0.12352"
    >
      Toast duration
    </label>
    <div
      class="void-select"
      id="0.12352"
    >
      <div
        class="void-option"
      >
        1s
      </div>
      <div
        class="void-option"
      >
        3s
      </div>
      <div
        class="void-option active"
      >
        5s
      </div>
      <div
        class="void-option"
      >
        10s
      </div>
    </div>
  </div>
  <button
    class="void-button"
  >
    Test Toasts
  </button>
</div>
`;
